atc/worker/placement.go

```diff
@@ -123,31 +194,27 @@ func (strategy *LimitActiveTasksPlacementStrategy) Choose(logger lager.Logger, w
 		}
 	}
 
-	leastBusyWorkers := workersByWork[minActiveTasks]
-	if len(leastBusyWorkers) < 1 {
-		return nil, nil
-	}
-	return leastBusyWorkers[strategy.rand.Intn(len(leastBusyWorkers))], nil
+	return workersByWork[minActiveTasks], nil
 }
 
-func (strategy *LimitActiveTasksPlacementStrategy) ModifiesActiveTasks() bool {
+func (strategy *LimitActiveTasksPlacementStrategyNode) ModifiesActiveTasks() bool {
 	return true
 }
 
-type RandomPlacementStrategy struct {
+type RandomPlacementStrategyNode struct {
 	rand *rand.Rand
 }
 
-func NewRandomPlacementStrategy() ContainerPlacementStrategy {
-	return &RandomPlacementStrategy{
+func newRandomPlacementStrategyNode() ContainerPlacementStrategyChainNode {
+	return &RandomPlacementStrategyNode{
 		rand: rand.New(rand.NewSource(time.Now().UnixNano())),
 	}
 }
 
-func (strategy *RandomPlacementStrategy) Choose(logger lager.Logger, workers []Worker, spec ContainerSpec) (Worker, error) {
-	return workers[strategy.rand.Intn(len(workers))], nil
+func (strategy *RandomPlacementStrategyNode) Choose(logger lager.Logger, workers []Worker, spec ContainerSpec) ([]Worker, error) {
+	return []Worker{workers[strategy.rand.Intn(len(workers))]}, nil
```

Comment:
Perhaps this should return all workers and let `containerPlacementStrategy.Choose()` deal with selecting at random.

Though there's no good reason to do it, if you specified `CONCOURSE_CONTAINER_PLACEMENT_STRATEGY= random,volume-locality`, the `volume-locality` strategy would have no effect, since we already filtered the list down to a single worker.

So, `random` is in a bit of a weird spot. It might be more valuable to view these as `XYZPlacementStrategyNodes` as filters - in which case, `RandomPlacementStrategyNode` is really an `IdentityFilter` or something like that - i.e. it just returns the full list of workers that it's provided.

Reply:
This makes sense. I thought to keep `random` strategy for backward-compatibility. Now, I think we can just deprecate `random` strategy, and states that in the option description.

Reply:
I don't know if we should get rid of random altogether - it just doesn't totally make sense with chaining. I think just making the RandomPlacementStrategyNode return the full list of workers is sufficient, though, since it won't lead to any unexpected behaviour with chaining

Reply:
I don't remove "random" in this PR. I just added option description to state deprecation.